### ISSUE TITLE
how to make multiple db session

### ISSUE THREAD

**Issue by shaidiren** [siddhantgoel/tornado-sqlalchemy#6]
Normally, we need more than one database session in models, how to deal with this situation?

**Comment by siddhantgoel** [comment]
Good question. I could imagine the API supporting multiple database sessions at some point, identified by names (or some other way of mapping a string to a session factory). At the moment though it's not supported. But I'm open to discussions and/or pull requests, in case you already have ideas on how to go about doing it.

**Comment by siddhantgoel** [comment]
On second thoughts, @shaidiren do you mean multiple sessions for the same handler or multiple sessions each pointing to a different database URL?

**Comment by weishen** [comment]
I have a need of multiple sessions for the same handler, meaning in the handler I have to use different databases.

**Comment by dingyaguang117** [comment]
how about this?

```
class NativeCoroutinesRequestHandler(SessionMixin, RequestHandler):
    async def get(self):
        with self.make_session('foo') as session:
            count = await as_future(session.query(UserModel).count)

        self.write('{} users so far!'.format(count))

app = Application(
   handlers,
   session_factory={
        'default': make_session_factory('postgres://user:password@host/database'),
        'foo': make_session_factory('postgres://user:password@host_foo/database'),
   }
)
```

**Comment by siddhantgoel** [comment]
In that case `self.session` would be ambiguous.

**Comment by dingyaguang117** [comment]
> In that case `self.session` would be ambiguous.

We can create a ` self.sessions('foo')`

and self.session is an alias for `self.sessions('default')`

**Comment by siddhantgoel** [comment]
Hmm good point. Let me play around a little bit with this idea locally and get back. :)

**Comment by dingyaguang117** [comment]
To support multiple databases, we have to solve a series problems

like declarative_base

```
# we have to use original sqlalchemy.ext.declarative.declarative_base 
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()
BaseFoo = declarative_base()

class User(Base):
    __tablename__ = 'users'

    id = Column(BigInteger, primary_key=True)
    username = Column(String(64), unique=True)

    def __init__(self, username):
        self.username = username

class Log(BaseFoo):
    __tablename__ = 'logs'

    id = Column(BigInteger, primary_key=True)
    title = Column(String(64))
    content = Column(String(64))

    def __init__(self, title, content):
        self.title = title
        self.content = content

```

**Comment by dingyaguang117** [comment]
@siddhantgoel 
hi, I'm trying to implement the feature, but  my modification failed to pass your test case,

```
    def test_distinct_sessions(self):
        sessions = set()

        class Handler(SessionMixin):
            def __init__(h_self):
                h_self.application = Mock()
                h_self.application.settings = {'session_factory': self.factory}

            def run(h_self):
                session = h_self.session

                sessions.add(id(session))
                value = session.query(User).count()

                session.commit()
                session.close()

                return value

        Handler().run()
        Handler().run()

        self.assertEqual(len(sessions), 2)

```
[My modification](https://github.com/dingyaguang117/tornado-sqlalchemy/blob/f2179624c16717c4519837c671dcab02cd1bf7f0/tornado_sqlalchemy/__init__.py#L90)

I found the reason, But I can't understand the intention of this test.

```
class SessionMixin:
    _sessions = {}  # it will be the same one in all instances
    application = None  # type: Optional[Application]
```

Is the reason for this thread safety?

**Comment by siddhantgoel** [comment]
In this case I would store the `_sessions` variable somewhere else instead of in the mixin.

**Comment by siddhantgoel** [comment]
But the multiple `declarative_base` instances is a bigger issue that I see here.

**Comment by dingyaguang117** [comment]
> In this case I would store the `_sessions` variable somewhere else instead of in the mixin.

@siddhantgoel 
You are right!

I put it in initialize, and it works well.

```
class SessionMixin:
    application = None  # type: Optional[Application]

    def initialize(self):
        self._sessions = {}
```

**Comment by dingyaguang117** [comment]
> But the multiple `declarative_base` instances is a bigger issue that I see here.

Yeah, I made a [modification](https://github.com/dingyaguang117/tornado-sqlalchemy/blob/dd553b1ffb0a0fc6ab92d7fda8d5536a72caa574/tornado_sqlalchemy/__init__.py#L167) to declarative_base

```
class _declarative_base:
    def __init__(self):
        self._bases = {}

    def __call__(self, db='default', **kwargs):
        base = self._bases.get(db)
        if not base:
            base = sa_declarative_base(**kwargs)
            base._Session = sessionmaker()
            self._bases[db] = base
        return base

declarative_base = _declarative_base()
```

and it works like [this](https://github.com/dingyaguang117/tornado-sqlalchemy/blob/dd553b1ffb0a0fc6ab92d7fda8d5536a72caa574/tests/_common.py#L59): 


```
Base = declarative_base()
BaseFoo = declarative_base(db='foo')
BaseBar = declarative_base(db='bar')

class User(Base):
    __tablename__ = 'users'

    id = Column(BigInteger, primary_key=True)
    username = Column(String(64), unique=True)

    def __init__(self, username):
        self.username = username

class Foo(BaseFoo):
    __tablename__ = 'foo'

    id = Column(BigInteger, primary_key=True)
    foo = Column(String(64))

    def __init__(self, foo):
        self.foo = foo

Base.metadata.create_all(self.factories['default'].engine)
BaseFoo.metadata.create_all(self.factories['foo'].engine)
```
but it's a little tedious, do you have any suggestions？

**Comment by siddhantgoel** [comment]
Yep, it does feel tedious. And also, what happens when you use the wrong `Base` and `factory` combination. That's also a potential error waiting to happen.

I'd like to think about a solution here. If you have any other ideas on this, please feel free to comment. Ideally let's try to fix the API first before any implementation happens.

**Comment by dingyaguang117** [comment]
You are right, we need an elegant API first

**Comment by dingyaguang117** [comment]
Another idea is defining `__database__` in the Model

```
Base = declarative_base()

class User(Base):
    __database__ = 'db_1'
    __tablename__ = 'users'

    id = Column(BigInteger, primary_key=True)
    username = Column(String(64), unique=True)

    def __init__(self, username):
        self.username = username
```

but how should we create session?

```
self.make_session_for_model(User) 
```
?

**Comment by dingyaguang117** [comment]
flask_sqlalchemy  create a query object for every Model. and the usage looks like:

```
User.query.fitler(User.id==1).first()
```

and the session is managed implicitly. 

But I don't  know how to implement it.

**Comment by dingyaguang117** [comment]
@siddhantgoel 
How about this:

create an init_sa function to bind engine to Base, and the create_all and drop_all would be shorten to

```
init_sa(application)
Base.metadata.create_all()
BaseFoo.metadata.create_all()
BaseBar.metadata.create_all()
```

the init_sa should be :
```

def init_sa(application):
    """
      set Base.metadata.bind with the corresponding engine
      And it's a chance to add features to Models
    """

    factory = application.settings.get('session_factory')

    if not isinstance(factory, dict):
        factories = {'default': factory}
    else:
        factories = factory

    for db, base in declarative_base._bases.items():
        factory = factories.get(db)
        if not factory:
            raise MissingFactoryError('No factory {} in application settings'.format(db))
        base.__database__ = db

        # [key line] equal to declarative_base(bind=factory.engine) when create Base
        base.metadata.bind = factory.engine
```

**Comment by siddhantgoel** [comment]
Since you referenced `flask_sqlalchemy`, I was reading through [how they manage multiple databases](https://flask-sqlalchemy.palletsprojects.com/en/2.x/binds/), and I somehow really like how they do it.

One proposal along those lines would be to require the user to only provide either one of `sqlalchemy_database_url` or `sqlalchemy_binds`. Something like this -

```python
# single
app = Application(
  handlers, sqlalchemy_database_url='postgresql://user:pass@localhost/db'
)

# multiple
app = Application(
  handlers, sqlalchemy_binds={
    'first': 'postgresql://user:pass@localhost/first',
    'second': 'postgresql://user:pass@localhost/second',
  }
)
```

... which would mean that the `SessionFactory` class is no longer useful and can be removed. The logic to select the correct session would then all be handled in `SessionMixin`.

Additionally, the `declarative_base` function could be replaced with a `make_declarative_base` function that takes an optional `key` parameter to distinguish between multiple bases.

This should result in the least amount of breakage since the upgrade path would be pretty clear (replace `session_factory: make_session_factory("...")` with `sqlalchemy_database_url: "..."`, and `declarative_base` with `make_declarative_base`).

Thoughts?

**Comment by dingyaguang117** [comment]
Great! Love it!

Anything else to consider？ Like: 

how to avoid using a wrong database and forgot which database the table is in.

**Comment by siddhantgoel** [comment]
`make_declarative_base` can return multiple bases, which the tables than can use. This should take care of the `create_all` case you mentioned in your earlier comment.

For querying, I don't think it's that easy since we're not explicitly keeping track of which bind contains which table. I have the feeling it'll make things a lot more complicated to work with. But open to suggestions.

**Comment by dingyaguang117** [comment]
@siddhantgoel  

flask_sqlalchemy's solution is really elegant.

Extending Session with  `__bind_key__`  supporting, and user don't need to make different sessions for different databases.

**Comment by siddhantgoel** [comment]
Yep. Let's think about `__bind_key__` in a later release and do things incrementally in the meanwhile.

Would you like to work on this or should I?

**Comment by dingyaguang117** [comment]
I'd like to do the work :)  But need your help in reviewing it.

**Comment by dingyaguang117** [comment]
@siddhantgoel 

For configuration parameter names, do you prefer uppercase or lowercase?

`sqlalchemy_database_url` or `SQLALCHEMY_DATABASE_URL `

**Comment by siddhantgoel** [comment]
> I'd like to do the work :) But need your help in reviewing it.

OK! :)

Please use lowercase for arguments being passed to the `Application` constructor.

So just to make sure we're on the same page, the following changes are proposed -

1. support `sqlalchemy_database_url` and `sqlalchemy_binds` arguments to the `Application` constructor
2. remove `SessionFactory` class
3. maintain a map of `sessionmaker` instances and `session`s inside the `SessionMixin` class
4. modify the `make_session` function to accept a `key` parameter which returns the requested session
5. return a `default` session when no `key` is passed
6. replace `declarative_base` with a `make_declarative_base` function which takes (optional) `key`

That's about it, right? Did I forget something?

**Comment by dingyaguang117** [comment]
After in-depth analysizing flask_sqlalchemy, I found the [sesionmaker](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.params.binds) supporting `binds` parameter which can manage the connections well.

The only thing is to create a [customized Session class](https://github.com/pallets/flask-sqlalchemy/blob/db02fd4c821b8aa8c6e47391fce6f064def6a363/flask_sqlalchemy/__init__.py#L144) which override the `get_bind` method.

And there will be only one session.

**Comment by siddhantgoel** [comment]
That sounds even better. Let's do it that way!

**Comment by dingyaguang117** [comment]
@siddhantgoel 

I [completed this feature](https://github.com/dingyaguang117/tornado-sqlalchemy/commit/435fc1456ba74bb2aeafa1d07eb73bfc519fe983), borrowed a lot from flask_sqlalchemy, 

Please review it and let me know if there is anything not good.

the example is [here](https://github.com/dingyaguang117/tornado-sqlalchemy/blob/master/examples/multi_databases.py)

**Comment by siddhantgoel** [comment]
Would you like to open a PR? I would then just add comments and provide feedback over there and we can resolve things one by one there.

**Comment by dingyaguang117** [comment]
sure!

**Comment by siddhantgoel** [comment]
I just published `v0.7.0` which includes this feature. Thanks @dingyaguang117 for your work on this!
